### ISSUE TITLE
'branch' variable can be referenced before assignment

### ISSUE THREAD

**Issue by sapols** [kLabUM/rrcf#56]
I was essentially replicating the [streaming example](https://github.com/kLabUM/rrcf/blob/master/docs/streaming.md) in this repo except with [my own dataset](https://github.com/sapols/Satellite-Telemetry-Anomaly-Detection/blob/master/Data/WheelTemperature.csv.zip) and the code broke with the error shown in the screenshot.

<img width="1103" alt="rrcf bug" src="https://user-images.githubusercontent.com/14846863/58190540-faac6580-7c79-11e9-93b5-3b4f354039c1.png">

My full code is [here](https://github.com/sapols/Satellite-Telemetry-Anomaly-Detection/blob/master/score_with_rrcf.py).

The bug is at line 460 of `rrcf.py`: `local variable 'branch' referenced before assignment`. What's going wrong is beyond me. Will someone please look into this? I work at LASP, a laboratory in Boulder, CO, and we're considering using this code in production, but we can't while this bug exists.

**Comment by mdbartos** [comment]
Hi @sapols 

Taking a look at this now.

I suspect that it's a bug with how duplicate points are handled.

Thanks,
MDB

**Comment by sapols** [comment]
Thank you for responding so quickly, MDB. Let me know if I can provide any more info or help in any way.

**Comment by mdbartos** [comment]
Ok. I believe I've found the reason the bug is being triggered.

It's happening when you try to insert a point that is a near-duplicate, but doesn't meet the tolerance criteria for a duplicate in at least one dimension. So, the point that it failed on for me was:

```
array([24.11841722, 24.11841722, 24.11841722, 24.11841722, 24.11841722,
       24.11841722, 24.11841722, 24.11841722, 24.11841722, 24.11841722,
       24.11841722, 24.11841722, 24.11841722, 24.11841722, 24.11841722,
       24.11841722, 24.11841722, 24.09805235])
```

The difference between this point and the nearest duplicate was:

```
array([ 0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        , -0.02036487])
```

So it wasn't identified as a duplicate, but was possibly too close to the nearest point to successfully find a cut. I will have to look into why the algorithm wasn't able to find a cut. Another issue is that it was trying to be inserted into a single-point tree (all duplicates), which might be triggering some kind of edge case.

**Comment by mdbartos** [comment]
I think I may have found the issue. In `forget_point`:

```python
        # If parent is the root...
        if parent is self.root:
            # Delete parent
            del parent
            # Set sibling as new root
            sibling.u = None
            if isinstance(sibling, Leaf):
                sibling.d = 0
            self.root = sibling
            # Update depths
            self.map_leaves(sibling, op=self._increment_depth, inc=-1)
            return self.leaves.pop(index)
```

Should be something like:

```python
        # If parent is the root...
        if parent is self.root:
            # Delete parent
            del parent
            # Set sibling as new root
            sibling.u = None
            self.root = sibling
            # Update depths
            if isinstance(sibling, Leaf):
                sibling.d = 0
            else:
                self.map_leaves(sibling, op=self._increment_depth, inc=-1)
            return self.leaves.pop(index)
```

Still need to test.

**Comment by mdbartos** [comment]
Hi @sapols,

I believe I fixed the issue. I've pulled the modified code into master. Still need to upload to pypi.

Let me know if you are still running into this issue.

A couple notes:

- If you're running on a large dataset like this one, I'd recommend using either:
  - Batch mode: https://klabum.github.io/rrcf/taxi.html
  - Parallelization: https://github.com/kLabUM/rrcf/issues/28

Also, instead of FIFO sampling, I'd probably recommend using something like reservoir sampling:
https://en.wikipedia.org/wiki/Reservoir_sampling

Thanks,
MDB
